### ISSUE TITLE
memcpy + context ambiguity

### ISSUE THREAD

**Issue by TApplencourt** [KhronosGroup/SYCL-Docs#184]
Hi,

Can one copy USM buffers that were created in different contexts using `memcpy` (example 1)? 

In another way. is this code sound?
```
queue Q1, Q2;
int *B1 = sycl::device_malloc<int>(Q1, N);
int *B2 = sycl::device_malloc<int>(Q2, N);
Q1.memcpy(B1,B2, N*sizeof(int))
```

#  Semantic

The memcopy documentation said:

>Copies numBytes of data from the pointer src to the pointer dest. Both dest and src may be either host or USM pointers. For more detail on USM, please see Section 4.8.

Where Section 4.8 famously said:

> USM allocations are only guaranteed to be valid in the SYCL context in which they were created. The behavior of accessing a USM allocation in a different context from which it was created is undefined. Attempting to access a kind of USM allocation that is not supported by the device results in undefined behavior.

`Section 4.8 ` makes the code show in Example 1 obviously undefined behavior (two contexts are used).  But because `memcpy` allows copy from any USM pointer to a host pointer,  it's trivial to implement a `context-independent memcopy`. 

The code below is sound in any spec interpretation I'm aware of:
```
queue Q1, Q2;
int *B1 = sycl::device_malloc<int>(Q1, N);
int *B2 = sycl::device_malloc<int>(Q2, N);
int *B = alloc(N*sizeof(int));

Q2.memcpy(B,B2, N*sizeof(int))
Q1.memcpy(B1,B, N*sizeof(int))
```
But you need to have access to both contexts in the API.

# TL;DR

The specification may want to clarify is a `sycl::queue::memcpy` between two USM Buffer bellowing to 2 different contexts is undefined behavior or not.   My understanding is that it's not possible, but I agree that one can read the "Both dest and src may be either host or USM pointers"  of the sycl::memcpy description in a little ambiguous maner.

This will help different compilers/backend to behave similarly, and for users to not rely on "subtle" undefined behavior.

(pinging @AerialMantis who is a far better writer than me. Gordon, please feel free to clarify/correct anything I wrote)

**Comment by gmlueck** [comment]
My notes from today's meeting:

* Clarify description of explicit copy APIs to say that any parameter that is a USM pointer must have been allocated from the same context as the `queue`.
* Clarify that for the explicit copy APIs that take both an `accessor` and also a pointer, the pointer must be accessible on the host.  This includes USM pointers, but not USM device pointers.

**Comment by keryell** [comment]
#### 2021/09/02 SYCL Khronos committee meeting
@AerialMantis describing since @TApplencourt not present:
When we have regular buffers, the SYCL runtime can manage the contexts in the background. But could the same thing works with USM?
Can the pointers be used in a single copy involving 2 different contexts or do we need to use 2 copies, each one with the right queue and passing through the host in the middle?
A user will prefer the single copy use-case and the implementation can use P2P if supported.
This would require the system to track the pointer allocation to work but actually today the runtime has to track the pointers anyway, or at least the lower runtime (like Level0, CUDA...).
This sounds difficult to have this to work for example between 2 GPUs from different vendors.
@jbrodman has pushed in Intel implementation the idea of a default platform context which would help. The fact we have so many contexts in SYCL is making SYCL overly complex. Often an OpenCL program just use 1 context...
We could remove the need for context at the SYCL level.
Contexts are dynamic objects and have states.
If we can have all the devices in the same context it solves the problem.
At some point we will need some API to test P2P capabilities between devices. On @jbrodman TODO-list.
We need some clarification.
Currently we say the pointers can be either a host or USM.
At least we should clarify that when there are 1 or 2 USM pointers involved they need to have been allocated in the same context related to the involved queue.
@gmlueck + @jbrodman + @AerialMantis to discuss together to converge on a PR.

**Comment by TApplencourt** [comment]
Thanks for the update and sorry that I missed the call...

I will talk about this topic during Monday's Talk at the summer session (https://www.khronos.org/events/sycl-summer-sessions-2021), I hope it will provide a good summary for a general audience. 

I have some bandwidth to help @gmlueck, @jbrodman, @AerialMantis craft a PR if they are interested.

**Comment by fraggamuffin** [comment]
can close based on GL's MR.

**Comment by gmlueck** [comment]
I think we should keep this open until #265 is merged.

BTW, I am still waiting for some more approvals on that PR ...

**Comment by TApplencourt** [comment]
Also opened #281, which should fix this issue if accepted.
